File: evil_factorial.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

const int MOD = 998244353;

// ฟังก์ชันสำหรับหา Power (x^y % p) ใช้สำหรับหา Modular Inverse
long long power(long long base, long long exp) {
    long long res = 1;
    base %= MOD;
    while (exp > 0) {
        if (exp % 2 == 1) res = (res * base) % MOD;
        base = (base * base) % MOD;
        exp /= 2;
    }
    return res;
}

// ฟังก์ชันหา Modular Inverse โดยใช้ Fermat's Little Theorem
long long modInverse(long long n) {
    return power(n, MOD - 2);
}

struct Query {
    int n, m, id;
};

int main() {
    // 1. Fast I/O: ปิด sync กับ stdio เพื่อให้อ่านเขียนไวที่สุด
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int Q;
    if (!(cin >> Q)) return 0;

    vector<Query> queries(Q);
    vector<int> checkpoints;
    
    // 0! = 1 เสมอ ใส่เป็นจุดตั้งต้น
    checkpoints.push_back(0); 

    // 2. รับ Input และเก็บจุดที่ต้องใช้ (n และ m)
    for (int i = 0; i < Q; i++) {
        cin >> queries[i].n >> queries[i].m;
        queries[i].id = i;
        checkpoints.push_back(queries[i].n);
        checkpoints.push_back(queries[i].m);
    }

    // 3. เรียงลำดับ Checkpoints และลบค่าซ้ำ
    sort(checkpoints.begin(), checkpoints.end());
    checkpoints.erase(unique(checkpoints.begin(), checkpoints.end()), checkpoints.end());

    // Map เพื่อเก็บค่า Factorial ของจุดที่ต้องการ (ใช้ Map ประหยัดเมมกว่า Array 10^8 ช่อง)
    // แต่ถ้า Q น้อยมาก (696) map ก็เร็วเพียงพอ ไม่ต้องใช้ Hash Table
    map<int, long long> fact_values;
    
    // กำหนดค่าเริ่มต้น
    long long current_fact = 1;
    int current_num = 0;
    fact_values[0] = 1;

    // 4. Sweep Line: วนลูปสร้าง Factorial รอบเดียว
    for (int point : checkpoints) {
        if (point == 0) continue;

        // คูณสะสมจากจุดเดิม ไปจนถึงจุดใหม่
        for (int i = current_num + 1; i <= point; i++) {
            current_fact = (current_fact * i) % MOD;
        }

        // บันทึกค่า
        fact_values[point] = current_fact;
        current_num = point;
    }

    // 5. คำนวณคำตอบแต่ละข้อ
    // n! / m! = n! * inv(m!)
    for (int i = 0; i < Q; i++) {
        long long fn = fact_values[queries[i].n];
        long long fm = fact_values[queries[i].m];
        
        long long inv_fm = modInverse(fm);
        long long ans = (fn * inv_fm) % MOD;

        cout << ans << "\n";
    }

    return 0;
}

File: evil_factorial.py
import sys


def solve():
    # 1. อ่านข้อมูลทั้งหมดรวดเดียวเพื่อความเร็ว (Fast I/O)
    input_data = sys.stdin.read().split()
    if not input_data:
        return

    iterator = iter(input_data)
    
    try:
        Q = int(next(iterator))
    except StopIteration:
        return

    queries = []
    needed_points = set()
    
    # 0! = 1 เป็นค่าพื้นฐานที่ต้องมี
    needed_points.add(0)

    # 2. เก็บคำถามทั้งหมด และระบุจุดที่ต้องการ (Events)
    for _ in range(Q):
        n = int(next(iterator))
        m = int(next(iterator))
        queries.append((n, m))
        needed_points.add(n)
        needed_points.add(m)

    # เรียงลำดับจุดที่ต้องการจากน้อยไปมาก
    sorted_points = sorted(list(needed_points))
    
    # 3. คำนวณ Factorial แบบรอบเดียว (Sweep Line)
    # เก็บค่า factorial ของจุดที่ต้องการไว้ใน dict เพื่อประหยัด Memory
    fact_map = {}
    current_fact = 1
    current_n = 0
    
    MOD = 998244353

    # วนลูปแค่ถึงจุดสูงสุดที่ต้องการ ไม่จำเป็นต้องถึง 10^8 ถ้าโจทย์ไม่ถาม
    for point in sorted_points:
        if point == 0:
            fact_map[0] = 1
            continue
        
        # คูณค่าต่อจากจุดเดิม ไปจนถึงจุดใหม่
        # การใช้ range ใน PyPy เร็วมาก
        temp = current_fact
        for i in range(current_n + 1, point + 1):
            temp = (temp * i) % MOD
        
        current_fact = temp
        fact_map[point] = current_fact
        current_n = point

    # 4. แสดงผลคำตอบ
    results = []
    for n, m in queries:
        # สูตร: (n! / m!) % p  =>  (n! * inverse(m!)) % p
        # Fermat's Little Theorem: inverse(a) = a^(p-2) % p
        
        val_n = fact_map[n]
        val_m = fact_map[m]
        
        # คำนวณ Modular Inverse
        inv_m = pow(val_m, MOD - 2, MOD)
        
        ans = (val_n * inv_m) % MOD
        results.append(str(ans))

    print('\n'.join(results))

if __name__ == '__main__':
    solve()